File: 백준/Silver/1874. 스택 수열/스택 수열.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main()
{
	stack<int> stack;
	int n,k,j =0,x =1;
	string array = "";
	cin >> n;

	for (int i = 1; i <= n; i++)  //n번 반복
	{
		cin >> k;  //수열 입력
		while(k >= x)   //4넣으면 4까지 push
		{
			stack.push(x++);
			array += '+';
		}
		if (stack.empty() || stack.top() < k) 
		{
		
			cout << "NO" << "\n";
			return 0;
		}
		if (k <= stack.top())
		{
			
			while (1)
			{

				if (k == stack.top())
				{
					array += '-';
					stack.pop();
					break;
				}
				else {
					array += '-';
					stack.pop();
				}
			}
			
			
		}
		
	}
	

	for (int i = 0; i < array.length(); i++)
	{
		if (i == array.length() - 1)
		{
			cout << array[i];
		}
		else {

			cout << array[i] << "\n";
		}
	}
	return 0;

}

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[]args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(br.readLine());
        int start = 1;
        Stack<Integer> stack = new Stack<>();
        for(int i  =0 ;i < n;i++){
            int tmp = Integer.parseInt(br.readLine());
            while(tmp >= start){//1부터 자리수 까지
                stack.push(start++);
                sb.append("+\n");
            }
            if(stack.empty() || stack.peek() < tmp){
                bw.write("NO\n");
                bw.flush();
                bw.close();
                return;
            }
            if(tmp <= stack.peek()){
                while(true){
                    if(tmp == stack.peek()){
                        sb.append("-\n");
                        stack.pop();
                        break;
                    }else{
                        sb.append("-\n");
                        stack.pop();
                    }
                }
            }
        }
        bw.write(sb.toString());


        bw.flush();
        bw.close();
    }





}